### PR TITLE
Cancel openstack's client.download

### LINKED ISSUE #379: How to cancel openstack's client.download?
I am new to `pkgcloud`. I do not know how to cancel client's download when request is closed or canceled.

My code is like below.

```js
var tempDownload = client.download({
	container: bucket,
	remote: object
});
tempDownload.pipe(response).on('finish', function() {
	console.log('object %s download finish.', object);
});
request.on('close', function() {
	tempDownload.unpipe(response);
//Need cancel client download here.
	console.log('request canceled');
})
```

Thank you for give advice.

-------
I think that maybe need to add `abort` function if someone need to abort.
Similar to `request` abort
https://github.com/request/request/pull/188

### PR DESCRIPTION
**Pull request by meteormatt**
1. Add `abort` on `openstack`'s client.download
1. `openstack`'s file request should able to add header such as `Range`
1. give `response` and `error` event to callback

close #379 also

### CODE REVIEW

**Comment by kenperkins** [comment]
Can you provide an example of how I would use this feature?

**Comment by meteormatt** [comment]
OK.

```
var pkgcloud = require('pkgcloud'),
    fs = require('fs');

var client = pkgcloud.storage.createClient({ /* ... */
});

var tempDownload = client.download({
    container: 'a-container',
    remote: 'remote-file-name.txt'
}).pipe(fs.createWriteStream('a-file.txt'));

setTimeout(function () {
    tempDownload.emit('abort');
}, 1000);
```

**Comment by kenperkins** [comment]
Awesome thanks!

**Comment by meteormatt** [comment]
When can it be merged?

**Comment by jcrugzz** [comment]
@kenperkins could the `abort` function be proxied here? seems cleaner to use a function call rather than emit an event.

**Comment by kenperkins** [comment]
@jcrugzz example pseudo code?

**Comment by jcrugzz** [comment]
@kenperkins Have the download stream itself have access to the abort function.

```js
var pkgcloud = require('pkgcloud'),
    fs = require('fs');

var client = pkgcloud.storage.createClient({ /* ... */
});

var tempDownload = client.download({
    container: 'a-container',
    remote: 'remote-file-name.txt'
}).pipe(fs.createWriteStream('a-file.txt'));

setTimeout(function () {
    tempDownload.abort();
}, 1000);
```

**Comment by kenperkins** [comment]
That's not bad. We could support both technically?

**Comment by jcrugzz** [comment]
@kenperkins yea its easy enough to support both, `abort()` function call just seems more intuitive to me :). @meteormatt good find, this functionality should definitely be exposed

**Comment by kenperkins** [comment]
@meteormatt if you can add the abort call we can merge today :smile:

**Comment by meteormatt** [comment]
I am looking at how to add the `abort` function at the `download` function
https://github.com/pkgcloud/pkgcloud/blob/master/lib/pkgcloud/openstack/storage/client/files.js#L176
I do not know how to do it yet.

**Comment by jcrugzz** [comment]
@meteormatt `abort()` is not accessible in that context, you have to add it to the `proxyStream` by "proxying" the `abort` function call of the `apiStream`. Ill show you in the code :)

File: lib/pkgcloud/openstack/client.js [hunk 1]
```diff
@@ -195,6 +195,10 @@ Client.prototype._request = function (options, callback) {
       self.emit('log::trace', 'Creating Authenticated Proxy Request');
       var apiStream = Client.super_.prototype._request.call(self, options, callback);
 
+      proxyStream.on('abort', function () {
+        apiStream.abort();
+      });
+
```
Comment (jcrugzz) [inline lib/pkgcloud/openstack/client.js]: @meteormatt here you could add the `abort()` function which will be on the returned `proxyStream` from `this._request` in `exports.download`.

```js
proxyStream.abort = function () {
  apiStream.abort();
});
```

**Comment by meteormatt** [comment]
@jcrugzz OK. Thank you very much. :relaxed: I know how to do now.

**Comment by meteormatt** [comment]
Is it OK to merge now?

**Comment by kenperkins** [comment]
Yep, I'll get to this today.

**Comment by kenperkins** [comment]
I'm also hoping to cut a release, so I'll do a lot of engineering after my morning meetings.

**Comment by meteormatt** [comment]
Thank you very much. :-)

在 2015/3/23 21:47, Ken Perkins 写道:

**Pull request merged**